source/h2_connection.c

```diff
@@ -521,13 +529,77 @@ static int s_decoder_on_ping(uint8_t opaque_data[AWS_H2_PING_DATA_SIZE], void *u
     }
 
     aws_h2_connection_enqueue_outgoing_frame(connection, ping_ack_frame);
-    s_try_write_outgoing_frames(connection);
     return AWS_OP_SUCCESS;
 error:
     CONNECTION_LOGF(ERROR, connection, "Ping ACK frame failed to be sent, error %s", aws_error_name(aws_last_error()));
     return AWS_OP_ERR;
 }
 
+static void s_aws_h2_decoder_change_settings(struct aws_h2_connection *connection) {
+    struct aws_h2_decoder *decoder = connection->thread_data.decoder;
+    uint32_t *settings_self = connection->thread_data.settings_self;
+    aws_h2_decoder_set_setting_header_table_size(decoder, settings_self[AWS_H2_SETTINGS_HEADER_TABLE_SIZE]);
+    aws_h2_decoder_set_setting_enable_push(decoder, settings_self[AWS_H2_SETTINGS_ENABLE_PUSH]);
+    aws_h2_decoder_set_setting_max_frame_size(decoder, settings_self[AWS_H2_SETTINGS_MAX_FRAME_SIZE]);
+}
+
+static void s_aws_h2_connection_change_peer_settings(
+    struct aws_h2_connection *connection,
+    struct aws_h2_frame_setting setting) {
+    uint32_t *settings_peer = connection->thread_data.settings_peer;
+    settings_peer[setting.id] = setting.value;
+}
+
+static int s_decoder_on_settings(struct aws_array_list *settings_array, size_t num_settings, void *userdata) {
+    struct aws_h2_connection *connection = userdata;
+    /* Once all values have been processed, the recipient MUST immediately emit a SETTINGS frame with the ACK flag
+     * set.(RFC-7540 6.5.3) */
+    CONNECTION_LOG(TRACE, connection, "Setting frame processing ends");
+    struct aws_h2_frame *settings_ack_frame = aws_h2_frame_new_settings(connection->base.alloc, NULL, 0, true);
+    if (!settings_ack_frame) {
+        CONNECTION_LOGF(
+            ERROR, connection, "Settings ACK frame failed to be sent, error %s", aws_error_name(aws_last_error()));
+        goto error;
+    }
+    aws_h2_connection_enqueue_outgoing_frame(connection, settings_ack_frame);
+    /* Store the change to encoder and connection after enqueue the setting ACK frame */
+    struct aws_h2_frame_setting setting;
+    struct aws_h2_frame_encoder *encoder = &connection->thread_data.encoder;
+    for (size_t i = 0; i < num_settings; i++) {
+        if (aws_array_list_get_at(settings_array, &setting, i)) {
+            CONNECTION_LOGF(
+                ERROR,
+                connection,
+                "Getting setting form array list failed, error %s",
+                aws_error_name(aws_last_error()));
+            goto error;
+        }
+        if (setting.id == AWS_H2_SETTINGS_HEADER_TABLE_SIZE) {
```

Comment:
trivial/efficiency: maybe just `continue` if new value matches existing value

Reply:
Same as the switch statement right?

Reply:
like, before the switch we can just do:
```
if (connection->thread_data.settings_peer[setting.id] == setting.value) {
    /* No change, don't do any work */
    continue;
}
switch (setting.id)
    ...
```